pkg/template/processor_test.go

```diff
@@ -0,0 +1,604 @@
+		})
+	})
+
+	t.Run("filter other resources", func(t *testing.T) {
+
+		t.Run("return one", func(t *testing.T) {
+			// given
+			objs := []runtime.RawExtension{
+				{
+					Object: ns1,
+				},
+				{
+					Object: rb1,
+				},
+			}
+			// when
+			result := template.Filter(objs, template.RetainAllButNamespaces)
+			// then
+			require.Len(t, result, 1)
+			assert.Equal(t, rb1, result[0].Object)
+
+		})
+		t.Run("return multiple", func(t *testing.T) {
+			// given
+			objs := []runtime.RawExtension{
+				{
+					Object: ns1,
+				},
+				{
+					Object: rb1,
+				},
+				{
+					Object: ns2,
+				},
+				{
+					Object: rb2,
+				},
+			}
+			// when
+			result := template.Filter(objs, template.RetainAllButNamespaces)
+			// then
+			require.Len(t, result, 2)
+			assert.Equal(t, rb1, result[0].Object)
+			assert.Equal(t, rb2, result[1].Object)
+		})
+
+		t.Run("return none", func(t *testing.T) {
+			// given
+			objs := []runtime.RawExtension{
+				{
+					Object: ns1,
+				},
+				{
+					Object: ns2,
+				},
+			}
+			// when
+			result := template.Filter(objs, template.RetainAllButNamespaces)
+			// then
+			require.Empty(t, result)
+		})
+	})
+}
+
+func TestProcessAndApply(t *testing.T) {
+
+	t.Run("should create namespace alone", func(t *testing.T) {
+
+		// given
+		s := addToScheme(t)
+		namespace, commit, user := templateVars()
+		values := paramsKeyValues(namespace, commit, user)
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s)
+
+		// when
+		objs, err := p.Process(templateContent(namespaceObj), values)
+		require.NoError(t, err)
+		err = p.Apply(objs, 2*time.Second)
+
+		// then
+		require.NoError(t, err)
+		verifyNamespace(t, cl, namespace)
+	})
+
+	t.Run("should create role binding alone", func(t *testing.T) {
+		// given
+		s := addToScheme(t)
+		namespace, commit, user := templateVars()
+		values := paramsKeyValues(namespace, commit, user)
+
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s)
+
+		// when
+		objs, err := p.Process(templateContent(roleBindingObj), values)
+		require.NoError(t, err)
+		err = p.Apply(objs, 2*time.Second)
+
+		// then
+		require.NoError(t, err)
+		verifyRoleBinding(t, cl, namespace)
+	})
+
+	t.Run("should create namespace and role binding", func(t *testing.T) {
+
+		t.Run("success", func(t *testing.T) {
+			// given
+			s := addToScheme(t)
+			namespace, commit, user := templateVars()
+			values := paramsKeyValues(namespace, commit, user)
+			cl := test.NewFakeClient(t)
+			p := template.NewProcessor(cl, s)
+
+			// when
+			objs, err := p.Process(templateContent(namespaceObj, roleBindingObj), values)
+			require.NoError(t, err)
+			err = p.Apply(objs, 2*time.Second)
+
+			// then
+			require.NoError(t, err)
+			verifyNamespace(t, cl, namespace)
+			verifyRoleBinding(t, cl, namespace)
+		})
+
+		t.Run("failure", func(t *testing.T) {
+
+			t.Run("client error", func(t *testing.T) {
+
+				t.Run("namespace does not exist", func(t *testing.T) {
```

Comment:
may I know why so many subtests and just not a longer subtest name such as:
```go
t.Run("apply should fail because of missing namespace", func(t *testing.T) {
```

Reply:
done (lost track of the commit, though :/ )

Reply:
my bad. Done in 18dcc52